feat(reposerver): pass otel tracing propagation carrier via env (#23988)

zzzz465 opened the pull request:
<!--
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.
-->

with argocd plugin feature, arbitrary code can be executed, which can also make HTTP requests. Since ArgoCD already supports OpenTelemetry tracing, if tracing context is passed to plugin, debugging would be much easier.

this PR makes ArgoCD RepoServer pass propagation data to the plugin using environment variables, following opentelemetry specification: https://opentelemetry.io/docs/specs/otel/context/env-carriers/

this Closes #23988 

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Title of the PR](https://argo-cd.readthedocs.io/en/latest/developer-guide/submit-your-pr/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

<!-- Please see [Contribution FAQs](https://argo-cd.readthedocs.io/en/latest/developer-guide/faq/) if you have questions about your pull-request. -->

zzzz465 committed:
pass context and fill trace context passing env when setup plugin

Signed-off-by: jungooji <zzzz465@naver.com>

zzzz465 committed:
use map[string]string instead of (map[string]string, bool)

Signed-off-by: jungooji <zzzz465@naver.com>

agaudreault reviewed (changes requested):
Also missing unit test to validate that the plugin receive tracing envs from context

zzzz465 committed:
Update reposerver/repository/repository.go

Co-authored-by: Alexandre Gaudreault <alexandre_gaudreault@intuit.com>
Signed-off-by: jungooji <zzzz465@naver.com>

zzzz465 committed:
revert

Signed-off-by: jungooji <zzzz465@naver.com>

zzzz465 committed:
align with existing codebase

Signed-off-by: jungooji <zzzz465@naver.com>

zzzz465 committed:
add unit test

Signed-off-by: jungooji <zzzz465@naver.com>

zzzz465 committed:
go mod tidy

Signed-off-by: jungooji <zzzz465@naver.com>

zzzz465 committed:
support baggage

Signed-off-by: jungooji <zzzz465@naver.com>

zzzz465 commented:
> Also missing unit test to validate that the plugin receive tracing envs from context

I added missing unit test, and also changed how envs are passed. to minimize code diff, now envs are inserted in `getPluginEnvs` which is only called when GenerateManifest + CMP flow. 

also during fixing, I noticed that only TraceContext is propagated, baggage isn't. unless this is intentional, supporting both would be nice.

zzzz465 commented on reposerver/repository/repository.go:
```diff
@@ -1593,6 +1595,14 @@ func newEnv(q *apiclient.ManifestRequest, revision string) *v1alpha1.Env {
+
+	if traceEnv := traceutil.GetTraceEnvFromContext(ctx); len(traceEnv) > 0 {
+		for key, value := range traceEnv {
+			*env = append(*env, &v1alpha1.EnvEntry{Name: key, Value: value})
```
I changed the place where envs inserted instead.

zzzz465 committed:
fix after CI review

Signed-off-by: jungooji <zzzz465@naver.com>

zzzz465 committed:
fix test ordernig

Signed-off-by: jungooji <zzzz465@naver.com>

zzzz465 committed:
fix test ordernig

Signed-off-by: jungooji <zzzz465@naver.com>

zzzz465 committed:
fix ci lint

Signed-off-by: jungooji <zzzz465@naver.com>
